Update TLS Testing

jhughesbiot opened the pull request:
The current TLS testing does not actually test the TLS version indicated:
https://github.com/faucetsdn/daq/issues/213

This PR adds specific TLS version testing for 1.2 and 1.3.  At this time, it will pass an x509 cert test if either version passes and individually pass/fail/skip the specific TLS version tests.  Because a device won't typically support more than 1 version, passing of 1.2 and not 1.3 or vice versa, this will fail the module. 

Example Report Output:

> |pass|security.tls.v1.2|Other|Other||
|fail|security.tls.v1.3|Other|Other||
|pass|security.x509|Other|Other||

Looking for suggestions on best way to modify this addition to have the module pass.
Some scenarios that could work:
1. Test high to low, stop at first version passed, skip lower version tests and report higher versions that failed as skip.  The report log would shed more light on the actual failure causes but the module would still pass in this manner.
2. Add highest version required test parameter to be supplied from module_config.json from the device.
3. Change how the module itself determines success/failure and allow all versions to be tested as this PR currently does and pass if at least 1.2(default version, make configurable) passes. 

I believe 3 is the most comprehensive and valid of the tests but the validation of the module itself is not something I've been able to determine.

grafnu commented:
I took a quick look and don't see anything wrong with the code/changes, but I didn't form an opinion about the fundamental of v1.2/v1.3 and how it's handled -- hoping @pisuke can take a stab at that. Not really a technical question, but about policy.

grafnu commented:
Also, the tests are failing because you changed the tests output and so also need to update docs/device_report.md to match expectations!

jhughesbiot commented:
I agree my questions were more about the testing policy here more so than technical ones.

As far as the testing failure goes, the docs/device_report.md has an effect on the test results?  I thought that was just an example of a report.  I was under the impression, based on the example I showed, that because I now run two TLS version tests, that unless the device supported both TLS 1.2 and 1.3 that there would be a _RESULT fail_ in the module report and that was the cause.  Is that not how the module failure is determined?

grafnu commented:
The tests basically compare the (redacted e.g. dates) results with
docs/device_report.md -- so it's an example, yes, but also a *golden* example
to detect failure. So, whatever result is output by for device-02 (I think)
should be reflected there. It's not comprehensive by any means, but it does
provide a first-order mechanism for detecting when something goes awry. It
doesn't matter exactly what's in the file (pass/fail/skip/whatever)... as
long as it's consistent.


On Wed, Dec 11, 2019 at 4:32 PM jhughesbiot <notifications@github.com>
wrote:

> I agree my questions were more about the testing policy here more so than
> technical ones.
>
> As far as the testing failure goes, the docs/device_report.md has an
> effect on the test results? I thought that was just an example of a report.
> I was under the impression, based on the example I showed, that because I
> now run two TLS version tests, that unless the device supported both TLS
> 1.2 and 1.3 that there would be a *RESULT fail* in the module report and
> that was the cause. Is that not how the module failure is determined?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafnu/daq/pull/269?email_source=notifications&email_token=AAIEPDY66TZIZWSDCRI35UDQYGBADA5CNFSM4JZUDXB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGVBRCA#issuecomment-564795528>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPD5CVYPTZQMODVH6H23QYGBADANCNFSM4JZUDXBQ>
> .
>

jhughesbiot commented:
I added the changes there and still the module fails.  And for clarity by fails, I mean everything runs and reports as expected but is included here at the end of the log.

`ERROR:runner:Test` failures: ['03:bacext:1', '03:switch:1', **'03:tls:1**']

So we may have been talking about two different module failure scenarios since that file seemed to have no effect on my testing.

To confirm this, I reverted the changes and made all failed TLS version tests report as skipped and the module passed since there was no failed test.  This seems to indicate my original question concerning the mechanism that looks at all the RESULT entries in the report still needs to be addressed based on what process of multiple TLS version testing is decided.

grafnu commented:
Ah, that error means that the Docker container (for TLS) errored out. Not a
test, per se, but the module that contains the test.

This is
subset/security/Dockerfile.test_tls
that runs
subset/security/test_tls
which calls
[ $(check_for_fails_in "${result_lines[@]}") == false ]
which checks the result file for the module. You can see the output of the
Dockerfile in the test run logs
<https://travis-ci.org/grafnu/daq/jobs/623828059> if you search for
tls01/activate.log -- looks like the x509 is failing so it fails the module.

Note that I think this logic is wrong (not your logic, but the logic in
test_tls)... the *module *(docker file) shouldn't fail if a line-item *test*
fails... but that's another issue for another time.





On Wed, Dec 11, 2019 at 9:26 PM jhughesbiot <notifications@github.com>
wrote:

> I added the changes there and still the module fails. And for clarity by
> fails, I mean everything runs and reports as expected but is included here
> at the end of the log.
>
> ERROR:runner:Test failures: ['03:bacext:1', '03:switch:1', *'03:tls:1*']
>
> So we may have been talking about two different module failure scenarios
> since that file seemed to have no effect on my testing.
>
> To confirm this, I reverted the changes and made all failed TLS version
> tests report as skipped and the module passed since there was no failed
> test. This seems to indicate my original question concerning the mechanism
> that looks at all the RESULT entries in the report still needs to be
> addressed based on what process of multiple TLS version testing is decided.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafnu/daq/pull/269?email_source=notifications&email_token=AAIEPD7IAMC7RS3KOQABQGDQYHDPZA5CNFSM4JZUDXB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGVQESQ#issuecomment-564855370>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPD3WVQ2JOXKYS4BVDRDQYHDPZANCNFSM4JZUDXBQ>
> .
>

jhughesbiot commented:
Yes, that was my assumption about the reason of the failure, any one line item test fails, the whole module fails.  Couldn't determine what process was actually reading the value created by this

> [ $(check_for_fails_in "${result_lines[@]}") == false ]

I imagine the real solution is to change the way this value is generated.  Probably should be checking for missing test results instead of the results themselves.  Alternatively, it could be checking for x509 failure always but then OR the TLS version successes so as long as one TLS version passes AND the x509 test passes, the module passes.  I can make an update either way just want to get consensus on how this module verification should look with multiple TLS version testing.

pisuke commented:
Hi Jon and Trevor,

Thanks for looking into this!

> 3. Change how the module itself determines success/failure and allow all versions to be tested as this PR currently does and pass if at least 1.2(default version, make configurable) passes.
> 
> I believe 3 is the most comprehensive and valid of the tests but the validation of the module itself is not something I've been able to determine.

I also think the test should be changed so that instead of checking "independently" TLS v1.2 vs v1.3 we do a single test that passes if at least a default version like 1.2 is provided.
It's similar to the way we've chosen to implement the port scanning in a single test with nmap rather than having individual tests for each port.

So yes, agreed to go for option 3. This means following up with Himawan and Clare to let them know about the change of approach, but I can do that.

Thanks again!

grafnu commented:
"Change how the module itself determines success/failure and allow all
versions to be tested as this PR currently does and pass if at least
1.2(default version, make configurable) passes"

The module (docker container) really doesn't pass/fail -- it either is
*success* or *error* -- which is different than tests which *pass/fail/skip*.
All the tests in a module/container can *fail* while the container still
has *success*. The container would *error* if there was some fundamental
error -- e.g. a missing binary or file-not-found.

I know the current setup is confusing and the terminology is not
consistent. But, better to think of the desired state and code to that
since we'll be moving in that direction. At one point there was a 1:1
relationship to modules and tests, but that changed and now there's a
1:many relationship -- with different semantics.


On Mon, Dec 16, 2019 at 3:12 PM Francesco Anselmo <notifications@github.com>
wrote:

> Hi Jon and Trevor,
>
> Thanks for looking into this!
>
>
>    1. Change how the module itself determines success/failure and allow
>    all versions to be tested as this PR currently does and pass if at least
>    1.2(default version, make configurable) passes.
>
> I believe 3 is the most comprehensive and valid of the tests but the
> validation of the module itself is not something I've been able to
> determine.
>
> I also think the test should be changed so that instead of checking
> "independently" TLS v1.2 vs v1.3 we do a single test that passes if at
> least a default version like 1.2 is provided.
> It's similar to the way we've chosen to implement the port scanning in a
> single test with nmap rather than having individual tests for each port.
>
> So yes, agreed to go for option 3. This means following up with Himawan
> and Clare to let them know about the change of approach, but I can do that.
>
> Thanks again!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafnu/daq/pull/269?email_source=notifications&email_token=AAIEPD63FRAN5FT2GVRVOXTQZADMBA5CNFSM4JZUDXB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHAOQXA#issuecomment-566290524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPD7EX37VDIODX2M5ELTQZADMBANCNFSM4JZUDXBQ>
> .
>

jhughesbiot commented:
Just for clarity, as I think were all on the same page but want to confirm, is that the current code that tests and reports on both TLS versions remain as is and only modify the module failure/success detection logic to determine if at least TLS 1.2 was successful.  X.509 success is implied by a successful TLS version test so that shouldn't need to be directly verified.

i.e. Change this section to look for TLS test success/failures instead of any generic failure.

> [ $(check_for_fails_in "${result_lines[@]}") == false ]

grafnu commented:
1. A module maps to a docker container, and should have "return" states
   of completed/error (not success/failure). The return state of a
   module/container should *not* depend on the pass/fail state of tests, as
   they are independent factors. All tests could fail and the module would
   still complete (w/o error). The line "[ $(check_for_fails_in
   "${result_lines[@]}") == false ]" is in error and should not be there at
   all.  It was something that crept in when I wasn't looking and I haven't
   had time to clean it up. Realistically you can do whatever you want with
   that line but just keep in mind it will be deleted in the future and so has
   no purpose you can rely on.
   2. Tests are individual line-items like "TLS v1.2". They are reported as
   pass/fail/skipped in the report.
   3. There is separate logic (outside of the modules and test line-items)
   that determines if the line-items tests are "acceptable" at a device
   holistic level.





On Tue, Dec 17, 2019 at 8:02 AM jhughesbiot <notifications@github.com>
wrote:

> Just for clarity, as I think were all on the same page but want to
> confirm, is that the current code that tests and reports on both TLS
> versions remain as is and only modify the module failure/success detection
> logic to determine if at least TLS 1.2 was successful. X.509 success is
> implied by a successful TLS version test so that shouldn't need to be
> directly verified.
>
> i.e. Change this section to look for TLS test success/failures instead of
> any generic failure.
>
> [ $(check_for_fails_in "${result_lines[@]}") == false ]
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafnu/daq/pull/269?email_source=notifications&email_token=AAIEPDZOYZLL7TT4ZD6ENHDQZDZZBA5CNFSM4JZUDXB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHC4MUA#issuecomment-566609488>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPD7S7KLG7EQEBL5J5H3QZDZZBANCNFSM4JZUDXBQ>
> .
>

jhughesbiot commented:
1. I can remove that line as part of this PR if it needs to be removed as it is already causing issues with the line item testing and module validation.
2. Currently already doing this so no work should be needed here
3. I believe this refers to the docs/device_report.md  that needs to be update?  If so, I can make the changes there and push (hopefully) a final update for this PR.

grafnu commented:
1. Yes, great.
   2. Beautiful, thanks.
   3. The device_report.md update is just a matter of testing and a "gold
   standard" for what tests are there.  I'm talking more about the underlying
   logic that translates to an overall device pass/fail -- which shouldn't be
   combined with TLS and this PR -- so nothing to be done there (other than
   making the docs/device_report.md file match appropriate).

Cheers,
  Trevor


On Tue, Dec 17, 2019 at 9:01 AM jhughesbiot <notifications@github.com>
wrote:

>
>    1. I can remove that line as part of this PR if it needs to be removed
>    as it is already causing issues with the line item testing and module
>    validation.
>    2. Currently already doing this so no work should be needed here
>    3. I believe this refers to the docs/device_report.md that needs to be
>    update? If so, I can make the changes there and push (hopefully) a final
>    update for this PR.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafnu/daq/pull/269?email_source=notifications&email_token=AAIEPDZG435XOD2ADHQQW2DQZEAV7A5CNFSM4JZUDXB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHDHJNA#issuecomment-566654132>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPD76YV6XTIIDU7CZ5U3QZEAV7ANCNFSM4JZUDXBQ>
> .
>

jhughesbiot commented:
All these updates should now be available for review. 

TLS module now passes even with individual TLS version failure results.

cmdrun.log output:

> INFO:config:Writing config to inst/run-port-03/nodes/tls03/tmp/module_config.json
> INFO:docker:Target port 3 test tls running
> INFO:docker:Target port 3 test tls passed 3.362936s

report results:
```
--------------------
security.tls.v1.2
--------------------
Verify the device supports TLS 1.2 (as a client)
--------------------
See log above
--------------------
RESULT **pass** security.tls.v1.2

--------------------
security.tls.v1.3
--------------------
Verify the device supports TLS 1.3 (as a client)
--------------------
See log above
--------------------
RESULT **fail** security.tls.v1.3
```

grafnu commented on subset/security/Dockerfile.test_tls:
```diff
@@ -1,9 +1,9 @@
 
 # Do this alone first so it can be re-used by other build files.
-RUN $AG update && $AG install openjdk-8-jre
+RUN $AG update && $AG install openjdk-11-jre
```
Is jdk-11 actually required?  Because of the way Docker works, this step then will no longer be optimized with other system.  Not fatal, but would unnecessarily increase build time.

jhughesbiot replied:
Yes, JDK 8 does not support TLS 1.3.

Reverted to jdk 8 to confirm this and reran tests so that I could recreate the error 

> Exception main:TLSv1.3 SSLContext not available

grafnu commented on docs/device_report.md:
```diff
@@ -264,16 +264,25 @@ RESULT skip protocol.bacnet.pic Bacnet device not found.
 
 --------------------
-security.tls.v3
+security.tls.v1.2
```
I think this might need to be security.tls.v1_2  -- since the dot character is used to subsection the various test categories.

jhughesbiot replied:
Updated in all locations where this format was referenced

```
--------------------
security.tls.v1_2
--------------------
Verify the device supports TLS 1.2 (as a client)
--------------------
See log above
--------------------
RESULT pass security.tls.v1_2

--------------------
security.tls.v1_3
--------------------
Verify the device supports TLS 1.3 (as a client)
--------------------
See log above
--------------------
RESULT fail security.tls.v1_3
```

grafnu commented:
Was this PR still active?  We should either finalize and submit or close it out.

jhughesbiot commented:
I need to merge in the latest changes which fix some of the flaky tests and update the test_aux.out file to match the changes to TLS results.  I believe the password test may still fail this due to it's inconsistent behavior so we'll see what happens after that.

grafnu commented:
Ok -- I'd like to either close this out until it's ready for a re-review,
or push it though.  Just don't like to have things lingering in the queue
(trying to have a better process around all this).  Also, if possibly could
you resubmit to the upstream faucetsdn/daq repo (rather than grafnu/daq) --
as part of the same process cleanup!

Cheers,
  Trevor


On Mon, Feb 24, 2020 at 8:23 AM jhughesbiot <notifications@github.com>
wrote:

> I need to merge in the latest changes which fix some of the flaky tests
> and update the test_aux.out file to match the changes to TLS results. I
> believe the password test may still fail this due to it's inconsistent
> behavior so we'll see what happens after that.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafnu/daq/pull/269?email_source=notifications&email_token=AAIEPD65PU3ZVQHV6HYASDDREPX7RA5CNFSM4JZUDXB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMYQK5I#issuecomment-590415221>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIEPD4SLHCNYYEXJJHQVG3REPX7RANCNFSM4JZUDXBQ>
> .
>
